app/src/main/java/com/vibro/testviewpager2/PdfRenderingEngine.kt

```diff
@@ -0,0 +1,291 @@
+package com.vibro.testviewpager2
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Paint
+import android.graphics.Rect
+import android.graphics.pdf.PdfDocument
+import android.net.Uri
+import android.util.Log
+import com.bumptech.glide.Glide
+import io.reactivex.Observable
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.lang.Exception
+import java.lang.IllegalArgumentException
+import java.util.*
+import kotlin.collections.ArrayList
+import kotlin.math.floor
+
+data class PageToSave(val index:Int, val bitmap: Bitmap, val width:Int, val height:Int)
+
+class PdfRenderingEngine(private val context: Context, private val snRenderer: SnRenderer) :
+    RenderPagesProvider {
+
+    private val TAG = this.javaClass.simpleName
+
+    private val cache = PagesCache(this)
+    private val originalPages: MutableList<PageInfo> = mutableListOf()
+    private val currentPages: MutableList<PageInfo> by lazy {
+        hasChanges = true
+        ArrayList<PageInfo>(originalPages)
+    }
+
+    private var hasChanges = false
+    private var multifiles = false
+
+    fun open(files: List<File>): List<PageInfo> {
+        multifiles = files.size > 1
+        return snRenderer.open(files).apply { originalPages.addAll(this) }
+    }
+
+    fun close() {
+        snRenderer.close()
+        cache.clearCache()
+    }
+
+    @Throws(IllegalArgumentException::class, OperationNotSupported::class, NoChangesFound::class, IOException::class)
+    fun save(): Observable<String> {
+        if (multifiles) return Observable.error(OperationNotSupported())
+        if (!hasChanges) return Observable.error(NoChangesFound())
+
+        var counter = 0
+        val pdfDocument = PdfDocument()
+
+        fun providePageToSave(page: PageInfo): Observable<PageToSave> {
+            return if (page.pageChangesData == null) {
+                snRenderer.renderPage(page, SnRenderer.Quality.Normal)
+                    .map {
+                        val isPortrait = it.pageInfo.pageAttributes?.isPortrait()?:true
+                        val pageSize = it.pageInfo.pageAttributes?.pageSize
+                        if (it.bitmap == null || pageSize == null) throw IllegalArgumentException()
+                        val width = if (isPortrait) pageSize.first else pageSize.second
```

Comment:
Did you check this file on disk?
I suppose it will have small images in it because page size and width can be smaller then it displayed in SnRenderer